File: experiments/v4-gate2/tests/support/canonical-json.mjs
// Independent Node implementation of the frozen Gate 2 canonical JSON profile.

export const MIN_SAFE_INTEGER = -9007199254740991;
export const MAX_SAFE_INTEGER = 9007199254740991;

export class CanonicalizationError extends Error {
  constructor(reasonCode) {
    super(reasonCode);
    this.name = "CanonicalizationError";
    this.reasonCode = reasonCode;
  }
}

export function validateUnicodeScalars(value) {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new CanonicalizationError("CANONICAL_UNPAIRED_SURROGATE");
      }
      index += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      throw new CanonicalizationError("CANONICAL_UNPAIRED_SURROGATE");
    }
  }
}

function escapeString(value) {
  validateUnicodeScalars(value);
  const named = new Map([
    ['"', '\\"'],
    ["\\", "\\\\"],
    ["\b", "\\b"],
    ["\t", "\\t"],
    ["\n", "\\n"],
    ["\f", "\\f"],
    ["\r", "\\r"],
  ]);
  let output = '"';
  for (const scalar of value) {
    if (named.has(scalar)) {
      output += named.get(scalar);
    } else {
      const codePoint = scalar.codePointAt(0);
      output += codePoint <= 0x1f
        ? `\\u${codePoint.toString(16).padStart(4, "0")}`
        : scalar;
    }
  }
  return `${output}"`;
}

export function canonicalJson(value) {
  const active = new WeakSet();

  function encode(item) {
    if (item === null) return "null";
    if (item === true) return "true";
    if (item === false) return "false";
    if (typeof item === "number") {
      if (Object.is(item, -0)) {
        throw new CanonicalizationError("CANONICAL_NEGATIVE_ZERO");
      }
      if (!Number.isFinite(item)) {
        throw new CanonicalizationError("CANONICAL_NONFINITE");
      }
      if (!Number.isInteger(item)) {
        throw new CanonicalizationError("CANONICAL_FLOAT");
      }
      if (!Number.isSafeInteger(item)) {
        throw new CanonicalizationError("CANONICAL_INTEGER_OUT_OF_RANGE");
      }
      return String(item);
    }
    if (typeof item === "bigint") {
      throw new CanonicalizationError("CANONICAL_BIGINT");
    }
    if (typeof item === "string") return escapeString(item);
    if (Array.isArray(item)) {
      if (active.has(item)) throw new CanonicalizationError("CANONICAL_CYCLE");
      if (Object.getPrototypeOf(item) !== Array.prototype) {
        throw new CanonicalizationError("UNSUPPORTED_ARRAY_PROPERTY");
      }
      for (let index = 0; index < item.length; index += 1) {
        if (!Object.prototype.hasOwnProperty.call(item, index)) {
          throw new CanonicalizationError("SPARSE_ARRAY");
        }
      }
      const allowedKeys = new Set(["length", ...item.map((_, index) => String(index))]);
      if (Reflect.ownKeys(item).some((key) => typeof key !== "string" || !allowedKeys.has(key))) {
        throw new CanonicalizationError("UNSUPPORTED_ARRAY_PROPERTY");
      }
      active.add(item);
      try {
        return `[${item.map(encode).join(",")}]`;
      } finally {
        active.delete(item);
      }
    }
    if (typeof item === "object") {
      if (active.has(item)) throw new CanonicalizationError("CANONICAL_CYCLE");
      const prototype = Object.getPrototypeOf(item);
      if (prototype !== Object.prototype && prototype !== null) {
        throw new CanonicalizationError("CANONICAL_UNSUPPORTED_TYPE");
      }
      const keys = Object.keys(item);
      const ownKeys = Reflect.ownKeys(item);
      if (ownKeys.some((key) => typeof key !== "string" || !keys.includes(key))) {
        throw new CanonicalizationError("CANONICAL_UNSUPPORTED_OBJECT_PROPERTY");
      }
      for (const key in item) {
        if (!Object.prototype.hasOwnProperty.call(item, key)) {
          throw new CanonicalizationError("CANONICAL_UNSUPPORTED_OBJECT_PROPERTY");
        }
      }
      for (const key of keys) validateUnicodeScalars(key);
      keys.sort((left, right) => Buffer.compare(
        Buffer.from(left, "utf8"),
        Buffer.from(right, "utf8"),
      ));
      active.add(item);
      try {
        return `{${keys.map((key) => `${escapeString(key)}:${encode(item[key])}`).join(",")}}`;
      } finally {
        active.delete(item);
      }
    }
    throw new CanonicalizationError("CANONICAL_UNSUPPORTED_TYPE");
  }

  return encode(value);
}

export function canonicalBytes(value) {
  return Buffer.from(canonicalJson(value), "utf8");
}

export function validateAttemptId(value) {
  if (typeof value !== "number" || !Number.isInteger(value)) {
    throw new CanonicalizationError("ATTEMPT_ID_NOT_INTEGER");
  }
  if (Object.is(value, -0) || value === 0) {
    throw new CanonicalizationError(Object.is(value, -0) ? "CANONICAL_NEGATIVE_ZERO" : "ATTEMPT_ID_ZERO");
  }
  if (value < 0) throw new CanonicalizationError("ATTEMPT_ID_NEGATIVE");
  if (!Number.isSafeInteger(value)) {
    throw new CanonicalizationError("CANONICAL_INTEGER_OUT_OF_RANGE");
  }
  return value;
}

export function validateSafeInteger(value) {
  if (typeof value !== "number" || !Number.isInteger(value)) {
    throw new CanonicalizationError("CANONICAL_INTEGER_REQUIRED");
  }
  if (Object.is(value, -0)) throw new CanonicalizationError("CANONICAL_NEGATIVE_ZERO");
  if (!Number.isSafeInteger(value)) {
    throw new CanonicalizationError("CANONICAL_INTEGER_OUT_OF_RANGE");
  }
  return value;
}

export function validateDecimalMetadata(value, { allowNegative = false } = {}) {
  if (typeof value !== "string" || !/^-?(?:0|[1-9][0-9]*)$/.test(value)) {
    throw new CanonicalizationError("CANONICAL_DECIMAL_METADATA");
  }
  if (value.startsWith("-") && (!allowNegative || value === "-0")) {
    throw new CanonicalizationError("CANONICAL_DECIMAL_METADATA");
  }
  return value;
}

File: experiments/v4-gate2/tests/support/canonical_json.py
"""Independent Python implementation of the frozen Gate 2 canonical JSON profile."""

from __future__ import annotations

import math
import re
from typing import Any


MIN_SAFE_INTEGER = -9_007_199_254_740_991
MAX_SAFE_INTEGER = 9_007_199_254_740_991
DECIMAL_METADATA = re.compile(r"^-?(?:0|[1-9][0-9]*)$")


class CanonicalizationError(ValueError):
    def __init__(self, reason_code: str):
        super().__init__(reason_code)
        self.reason_code = reason_code


def _validate_string(value: str) -> None:
    try:
        value.encode("utf-8", errors="strict")
    except UnicodeEncodeError as error:
        raise CanonicalizationError("CANONICAL_UNPAIRED_SURROGATE") from error


def _escape_string(value: str) -> str:
    _validate_string(value)
    escapes = {
        '"': '\\"',
        "\\": "\\\\",
        "\b": "\\b",
        "\t": "\\t",
        "\n": "\\n",
        "\f": "\\f",
        "\r": "\\r",
    }
    result = ['"']
    for scalar in value:
        if scalar in escapes:
            result.append(escapes[scalar])
        elif ord(scalar) <= 0x1F:
            result.append(f"\\u{ord(scalar):04x}")
        else:
            result.append(scalar)
    result.append('"')
    return "".join(result)


def canonical_json(value: Any) -> str:
    active: set[int] = set()

    def encode(item: Any) -> str:
        if item is None:
            return "null"
        if item is True:
            return "true"
        if item is False:
            return "false"
        if isinstance(item, int):
            if item < MIN_SAFE_INTEGER or item > MAX_SAFE_INTEGER:
                raise CanonicalizationError("CANONICAL_INTEGER_OUT_OF_RANGE")
            return str(item)
        if isinstance(item, float):
            if not math.isfinite(item):
                raise CanonicalizationError("CANONICAL_NONFINITE")
            if item == 0 and math.copysign(1.0, item) < 0:
                raise CanonicalizationError("CANONICAL_NEGATIVE_ZERO")
            raise CanonicalizationError("CANONICAL_FLOAT")
        if isinstance(item, str):
            return _escape_string(item)
        if isinstance(item, list):
            if type(item) is not list:
                raise CanonicalizationError("UNSUPPORTED_ARRAY_PROPERTY")
            identity = id(item)
            if identity in active:
                raise CanonicalizationError("CANONICAL_CYCLE")
            active.add(identity)
            try:
                return "[" + ",".join(encode(child) for child in item) + "]"
            finally:
                active.remove(identity)
        if isinstance(item, dict):
            if type(item) is not dict:
                raise CanonicalizationError("CANONICAL_UNSUPPORTED_TYPE")
            identity = id(item)
            if identity in active:
                raise CanonicalizationError("CANONICAL_CYCLE")
            for key in item:
                if not isinstance(key, str):
                    raise CanonicalizationError("CANONICAL_NON_STRING_KEY")
                _validate_string(key)
            active.add(identity)
            try:
                keys = sorted(item, key=lambda key: key.encode("utf-8", errors="strict"))
                return "{" + ",".join(
                    _escape_string(key) + ":" + encode(item[key]) for key in keys
                ) + "}"
            finally:
                active.remove(identity)
        raise CanonicalizationError("CANONICAL_UNSUPPORTED_TYPE")

    return encode(value)


def canonical_bytes(value: Any) -> bytes:
    return canonical_json(value).encode("utf-8", errors="strict")


def validate_attempt_id(value: Any) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise CanonicalizationError("ATTEMPT_ID_NOT_INTEGER")
    if value == 0:
        raise CanonicalizationError("ATTEMPT_ID_ZERO")
    if value < 0:
        raise CanonicalizationError("ATTEMPT_ID_NEGATIVE")
    if value > MAX_SAFE_INTEGER:
        raise CanonicalizationError("CANONICAL_INTEGER_OUT_OF_RANGE")
    return value


def validate_decimal_metadata(value: str, allow_negative: bool = False) -> str:
    if not isinstance(value, str) or not DECIMAL_METADATA.fullmatch(value):
        raise CanonicalizationError("CANONICAL_DECIMAL_METADATA")
    if value.startswith("-") and (not allow_negative or value == "-0"):
        raise CanonicalizationError("CANONICAL_DECIMAL_METADATA")
    return value
